### PR TITLE
Feature/7232 multiple wrap components

### PR DESCRIPTION
**Pull request by DayS**
This PR fixes the issue #7232 : Adding multiple plugins with wrapComponents on the same core component seems broken

A new option `pluginsBehavior` has been added to keep backward compatibility for existing plugins.
When this flag is set to `chain`, the `system.combinePlugins()` method chain the plugins targeting the same core component, by providing the system to inner `systemExtend` call.

### Motivation and Context
I'm currently working on a set of plugins and stumble upon this issue as multiple of them needed to wrap `info` component.
Also, I think the project lacks some unique plug-points (like the `JumpToPath` which is used in multiple context).

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [x] Bug fixes (non-breaking change which fixes an issue)
- [ ] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [x] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [ ] My changes do not require a change to the project documentation.
- [x] My changes require a change to the project documentation.
- [x] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [x] My changes can and should be tested by unit and/or integration tests.
- [x] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by tim-lai** [comment]
@DayS Thanks for the PR, it looks good in general, and thanks for considering possibly breaking changes. A couple of feedback points:

- instead of `pluginsBehavior`, let's use `pluginsOptions: {}`. So then we would be checking for `pluginsOptions.pluginLoadType === 'chain'`
- add a default `pluginsBehavior` object to the `defaults` in `src/core/index.js`, as well as  `pluginsOptions.pluginLoadType: 'legacy'`
- add documentation for this new config option, and recommend that if one does NOT want the `chain` behavior, one should start adding `legacy` to their code now.

By adding `pluginsOptions`, we enable possible future options, while defining the default `pluginsOptions.pluginLoadType: legacy` gives indication that we are likely to change the default in the future. 

If you have any additional questions or thoughts, please feel free to share. Thanks again!

**Comment by DayS** [comment]
Hi, thanks for the review.

The `pluginsOptions` object seems like a better idea, I'll change that. Do we still need the `plugin` prefix on `pluginLoadType` ? It seems a little redundant.


And just to be sure :

> * add a default `pluginsBehavior`

You mean `pluginsOptions`, right ?

**Comment by tim-lai** [comment]
> The `pluginsOptions` object seems like a better idea, I'll change that. Do we still need the `plugin` prefix on `pluginLoadType` ? It seems a little redundant.
> 

I tend to favor verbosity. 😄 
 
> You mean `pluginsOptions`, right ?

correct. typo.

**Comment by DayS** [comment]
The PR has been updates with the changes. Let me know if it seems clear enough for the doc changes

File: src/core/system.js [hunk 1]
```diff
@@ -63,7 +64,7 @@ export default class Store {
   }
 
   register(plugins, rebuild=true) {
-    var pluginSystem = combinePlugins(plugins, this.getSystem())
+    var pluginSystem = combinePlugins(plugins, this.getSystem(), this.pluginsOptions.pluginLoadType)
```
Comment (tim-lai) [inline src/core/system.js:67]:
```suggestion
    var pluginSystem = combinePlugins(plugins, this.getSystem()
```

File: src/core/system.js [hunk 2]
```diff
@@ -35,6 +35,7 @@ export default class Store {
     deepExtend(this, {
       state: {},
       plugins: [],
+      pluginsOptions: {},
```
Comment (tim-lai) [inline src/core/system.js:38]:
```suggestion
```

Comment (tim-lai) [reply src/core/system.js:38]: along with previous suggested change, it should not be necessary to deepExtend the configuration object

Comment (tim-lai) [reply src/core/system.js:38]: hmm, nevermind

**Comment by tim-lai** [comment]
@DayS Whoops on my part, can you you revert commit 0fe6fbf "Apply suggestions from code review". Thanks.

File: src/core/system.js [hunk 3]
```diff
@@ -310,19 +311,21 @@ export default class Store {
 
 }
 
-function combinePlugins(plugins, toolbox) {
+function combinePlugins(plugins, toolbox, pluginLoadType) {
```
Comment (tim-lai) [inline src/core/system.js:314]: Ok, revised comment. we should receive the `pluginOptions` object here

File: src/core/system.js [hunk 4]
```diff
@@ -310,19 +311,21 @@ export default class Store {
...
     return assignDeep({}, plugins)
   }
 
   if(isFunc(plugins)) {
-    return combinePlugins(plugins(toolbox), toolbox)
+    return combinePlugins(plugins(toolbox), toolbox, pluginLoadType)
   }
 
   if(isArray(plugins)) {
+    const dest = pluginLoadType === "chain" ? toolbox.getComponents() : {}
```
Comment (tim-lai) [inline src/core/system.js:324]: revised comment... and then check for `pluginOptions.pluginLoadType` here

File: src/core/system.js (hunk 3 above)
Comment (tim-lai) [reply src/core/system.js:314]: where we pass in the `pluginsOption` in line 67 above

Comment (tim-lai) [reply src/core/system.js:314]: I also think we should be able to get `pluginsOptions` via `this.getConfigs().pluginsOptions`

File: src/core/system.js (hunk 2 above)
Comment (DayS) [reply src/core/system.js:38]: With further testing, I can confirm that we DO need to deepExtend the configuration object :)

File: src/core/system.js (hunk 4 above)
Comment (DayS) [reply src/core/system.js:324]: Ok, It may open some future tweaking this way

File: src/core/system.js (hunk 3 above)
Comment (DayS) [reply src/core/system.js:314]: > I also think we should be able to get `pluginsOptions` via `this.getConfigs().pluginsOptions`

It seems to be handled by this part : https://github.com/swagger-api/swagger-ui/blob/master/src/core/index.js#L163
Can you confirm ?

Comment (tim-lai) [reply src/core/system.js]: re: this.getConfigs(), I was thinking of something else. sorry for the confusion.

**Comment by tim-lai** [comment]
@DayS PR merged! Thanks for the collaboration and contribution!

**Comment by DayS** [comment]
Awesome !

You're welcome. I'm thrilling to see the next release and use this :)

**Pull request merged**